src/client/Client.cc

```diff
@@ -8905,26 +8888,16 @@ int Client::_read_async(Fh *f, uint64_t off, uint64_t len, bufferlist *bl)
                  << " max_periods=" << conf->client_readahead_max_periods << dendl;
 
   // read (and possibly block)
-  int r, rvalue = 0;
-  Mutex flock("Client::_read_async flock");
-  Cond cond;
-  bool done = false;
-  Context *onfinish = new C_SafeCond(&flock, &cond, &done, &rvalue);
+  int r = 0;
+  C_SaferCond onfinish("Client::_read_async flock");
```

Comment:
here too

Reply:
actually finisher calls complete(), which (by default) does finish(r) and then delete this, and C_SaferCond implements complete() instead of finish().  Should work I think!

Reply:
generally speaking trying to get away from ever calling finish()... everyone should call complete() instead so that the context can live off the heap

Reply:
have updated the code, some place have to new a C_SaferCond pointer, need to delete that pointer.

Reply:
Thanks for the info @liewegas !